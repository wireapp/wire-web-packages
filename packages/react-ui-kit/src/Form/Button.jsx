/*
 * Wire
 * Copyright (C) 2017 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import {COLOR} from '../Identity';
import PropTypes from 'prop-types';
import {Text} from '../Text';
import {defaultTransition} from '../Identity/motions';

const darkenAmount = 0.08;
const Button = Text.withComponent('button').extend`
  /* appearance */
  background-color: ${props => (props.disabled ? COLOR.DISABLED : props.backgroundColor)};
  border-radius: 8px;
  border: 0;
  cursor: ${props => (props.disabled ? 'default' : 'pointer')};
  display: inline-block;
  text-decoration: none;
  ${defaultTransition}

  /* positioning */
  height: 48px;
  line-height: 48px;
  max-width: 100%;
  ouline: none;
  padding: 0 32px;
  min-width: 184px;
  width: ${props => (props.block ? '100%' : 'auto')};
  &:hover,
  &:focus {
    /* appearance */
    text-decoration: none;
    background-color: ${props => (props.disabled ? COLOR.DISABLED : COLOR.shade(props.backgroundColor, darkenAmount))}
   }
`;

Button.propTypes = {
  ...Text.propTypes,
  backgroundColor: PropTypes.string,
  block: PropTypes.bool,
  disabled: PropTypes.bool,
  noCapital: PropTypes.bool,
};

Button.defaultProps = {
  ...Text.defaultProps,
  backgroundColor: COLOR.BLUE,
  block: false,
  bold: true,
  center: true,
  color: COLOR.WHITE,
  disabled: false,
  noCapital: false,
  nowrap: true,
  textTransform: 'uppercase',
  truncate: true,
};

const ButtonLink = Button.withComponent('a');

export {Button, ButtonLink};
