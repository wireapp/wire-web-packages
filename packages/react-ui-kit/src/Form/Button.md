Demo:

```js
import {Button, Container, Columns, Column, COLOR} from '@wireapp/react-ui-kit';

<Container>
  <Columns>
    <Column>Button</Column>
    <Column>
      <Button backgroundColor={COLOR.BLUE}>Button</Button>
    </Column>
  </Columns>
  <Columns>
    <Column>Disabled Button</Column>
    <Column>
      <Button disabled onClick={() => alert('This should not work')}>
        Disabled Button
      </Button>
    </Column>
  </Columns>
  <Columns>
    <Column>Loading Button</Column>
    <Column>
      <Button showLoading disabled onClick={() => alert('This should not work')}>
        Loading Button
      </Button>
    </Column>
  </Columns>
</Container>;
```
