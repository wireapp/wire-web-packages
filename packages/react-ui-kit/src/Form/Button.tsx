/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import styled from 'styled-components';
import {COLOR} from '../Identity';
import {defaultTransition} from '../Identity/motions';
import {Link, Text, TextProps} from '../Text';

interface ButtonProps extends TextProps {
  backgroundColor?: string;
  block?: boolean;
  disabled?: boolean;
  noCapital?: boolean;
}

type HTMLButtonProps = ButtonProps & React.ButtonHTMLAttributes<HTMLButtonElement>;

const darkenAmount = 0.06;
const Button = styled<HTMLButtonProps>(Text.withComponent(styled.button<HTMLButtonProps>``))<HTMLButtonProps>`
  background-color: ${props => (props.disabled ? COLOR.DISABLED : props.backgroundColor)};
  border-radius: 8px;
  border: 0;
  cursor: ${props => (props.disabled ? 'default' : 'pointer')};
  display: inline-block;
  text-decoration: none;
  margin-bottom: 16px;
  touch-action: manipulation;
  ${defaultTransition};
  height: 48px;
  line-height: 48px;
  max-width: 100%;
  outline: none;
  padding: 0 32px;
  min-width: 150px;
  width: ${props => (props.block ? '100%' : 'auto')};
  &:hover,
  &:focus {
    text-decoration: none;
    background-color: ${props => (props.disabled ? COLOR.DISABLED : COLOR.shade(props.backgroundColor, darkenAmount))};
  }
`;

Button.defaultProps = {
  backgroundColor: COLOR.BLUE,
  block: false,
  bold: true,
  center: true,
  color: COLOR.WHITE,
  disabled: false,
  fontSize: '16px',
  noCapital: false,
  noWrap: true,
  textTransform: 'uppercase',
  truncate: true,
};

const ButtonLink = styled(Button.withComponent(Link))`
  display: inline-block !important;
`;

ButtonLink.defaultProps = {
  ...Button.defaultProps,
};

export {Button, ButtonLink};
