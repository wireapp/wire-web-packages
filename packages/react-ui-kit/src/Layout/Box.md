Demo:

```js
import {Box} from '@wireapp/react-ui-kit';

<Box>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce a mattis nibh, sed maximus leo. Fusce a lacinia sem,
  vitae ornare dolor. Quisque rhoncus, magna non lacinia sagittis, erat augue fringilla metus, eu consectetur leo velit
  non lacus. Phasellus ipsum turpis, dapibus ut purus in, lobortis consectetur mi.
</Box>;
```
