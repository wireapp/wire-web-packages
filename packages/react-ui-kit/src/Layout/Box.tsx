/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

/** @jsx jsx */
import {ObjectInterpolation, jsx} from '@emotion/core';
import React from 'react';
import {COLOR} from '../Identity';

export interface BoxProps<T = HTMLDivElement> extends React.HTMLProps<T> {}

const boxStyles: (props: BoxProps) => ObjectInterpolation<undefined> = props => ({
  border: `2px solid ${COLOR.GRAY_LIGHTEN_72}`,
  borderradius: '8px',
  padding: '16px 32px',
});

const Box = (props: BoxProps) => <div css={boxStyles(props)} {...props} />;

export {Box, boxStyles};
