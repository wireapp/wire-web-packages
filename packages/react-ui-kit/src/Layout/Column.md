Demo:

```js
import {Container, Columns, Column, COLOR} from '@wireapp/react-ui-kit';

const ColumnsStyle = {
  marginBottom: '12px',
};

const ColumnStyle = {
  backgroundColor: COLOR.GRAY_LIGHTEN_32,
  border: `1px solid ${COLOR.GRAY_LIGHTEN_24}`,
};

<Container>
  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>
      <Columns>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
      </Columns>
    </Column>
    <Column style={ColumnStyle}>
      <Columns>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
      </Columns>
    </Column>
    <Column style={ColumnStyle}>
      <Columns>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
      </Columns>
    </Column>
  </Columns>
</Container>;
```
