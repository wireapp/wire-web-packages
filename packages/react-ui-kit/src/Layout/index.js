import setGlobalStyles from '../globalStyles';

setGlobalStyles();

export * from './Column';
export * from './Container';
export * from './Content';
export * from './Footer';
export * from './Header';
export * from './StyledApp';
