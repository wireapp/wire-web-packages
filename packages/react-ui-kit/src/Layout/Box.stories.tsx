/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import {Meta, StoryObj} from '@storybook/react';

import {Box} from './Box';

import {COLOR} from '../Identity/colors';
import {Text} from '../Text';

const meta: Meta<typeof Box> = {
  component: Box,
  title: 'Layout/Box',
  decorators: [
    Story => (
      <div style={{padding: '24px', maxWidth: '600px', margin: '0 auto'}}>
        <Story />
      </div>
    ),
  ],
};

export default meta;
type Story = StoryObj<typeof Box>;

const LOREM_IPSUM = `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce a mattis nibh, sed maximus leo. Fusce a lacinia sem,
vitae ornare dolor. Quisque rhoncus, magna non lacinia sagittis, erat augue fringilla metus, eu consectetur leo velit
non lacus. Phasellus ipsum turpis, dapibus ut purus in, lobortis consectetur mi.`;

export const Default: Story = {
  args: {
    children: LOREM_IPSUM,
  },
};

export const WithCustomPadding: Story = {
  args: {
    children: LOREM_IPSUM,
    style: {padding: '32px'},
  },
};

export const WithCustomBorder: Story = {
  args: {
    children: LOREM_IPSUM,
    style: {border: `2px solid ${COLOR.BLUE}`},
  },
};

export const WithFormattedContent: Story = {
  args: {
    children: (
      <>
        <Text block style={{fontSize: '20px', fontWeight: 'bold', marginBottom: '16px'}}>
          Important Notice
        </Text>
        <Text block style={{marginBottom: '12px'}}>
          {LOREM_IPSUM}
        </Text>
        <Text block style={{color: COLOR.GRAY}}>
          Last updated: 2025-01-01
        </Text>
      </>
    ),
  },
};

export const Nested: Story = {
  args: {
    children: (
      <>
        <Text block style={{marginBottom: '16px'}}>
          Outer box content
        </Text>
        <Box style={{backgroundColor: COLOR.GRAY_LIGHTEN_92}}>
          <Text>Inner box content</Text>
        </Box>
      </>
    ),
  },
};

export const Interactive: Story = {
  args: {
    children: (
      <div style={{textAlign: 'center'}}>
        <Text block style={{marginBottom: '16px'}}>
          Click the button below
        </Text>
        <button
          style={{
            backgroundColor: COLOR.BLUE,
            border: 'none',
            borderRadius: '4px',
            color: COLOR.WHITE,
            cursor: 'pointer',
            padding: '8px 16px',
          }}
        >
          Action Button
        </button>
      </div>
    ),
    style: {cursor: 'pointer'},
  },
};
