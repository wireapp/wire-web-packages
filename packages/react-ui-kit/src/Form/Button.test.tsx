/*
 * Wire
 * Copyright (C) 2019 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import React from 'react';
import {THEME_ID} from '../Layout';
import {matchComponent} from '../test/testUtil';
import {Button, ButtonLink} from './Button';

describe('"Button"', () => {
  it('renders', () => matchComponent(<Button>Submit</Button>));
  it('renders (dark theme)', () => matchComponent(<Button>Button</Button>, THEME_ID.DARK));
  it('renders when disabled', () => matchComponent(<Button disabled>Submit</Button>));
  it('renders as block', () => matchComponent(<Button block>Submit</Button>));
  it('renders in loading state', () => matchComponent(<Button showLoading>Submit</Button>));
});

describe('"ButtonLink"', () => {
  it('renders', () => matchComponent(<ButtonLink>Submit</ButtonLink>));
  it('renders (dark theme)', () => matchComponent(<ButtonLink>ButtonLink</ButtonLink>, THEME_ID.DARK));
  it('renders in loading state', () => matchComponent(<ButtonLink showLoading>Submit</ButtonLink>));
});
