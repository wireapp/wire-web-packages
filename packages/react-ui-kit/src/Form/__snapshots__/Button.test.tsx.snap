// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"Button" renders (dark theme) 1`] = `
.emotion-0 {
  color: #fff;
  display: inline-block;
  font-size: 16px;
  font-weight: 600;
  opacity: 1;
  overflow: hidden;
  text-align: center;
  text-overflow: ellipsis;
  text-transform: uppercase;
  white-space: nowrap;
  background-color: #0c78ba;
  border: 0;
  border-radius: 8px;
  cursor: pointer;
  height: 48px;
  line-height: 48px;
  margin-bottom: 16px;
  max-width: 100%;
  min-width: 150px;
  outline: none;
  padding: 0 32px;
  -webkit-text-decoration: none;
  text-decoration: none;
  touch-action: manipulation;
  -webkit-transition: all 0.24s;
  transition: all 0.24s;
  width: auto;
}

.emotion-0:hover,
.emotion-0:focus {
  background-color: rgb(11,113,175);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1 {
  background: #000;
  -webkit-transition: background 0.15s;
  transition: background 0.15s;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
  >
    Button
  </button>
</div>
`;

exports[`"Button" renders 1`] = `
.emotion-1 {
  background: rgb(247,248,250);
  -webkit-transition: background 0.15s;
  transition: background 0.15s;
}

.emotion-0 {
  color: #fff;
  display: inline-block;
  font-size: 16px;
  font-weight: 600;
  opacity: 1;
  overflow: hidden;
  text-align: center;
  text-overflow: ellipsis;
  text-transform: uppercase;
  white-space: nowrap;
  background-color: #0c78ba;
  border: 0;
  border-radius: 8px;
  cursor: pointer;
  height: 48px;
  line-height: 48px;
  margin-bottom: 16px;
  max-width: 100%;
  min-width: 150px;
  outline: none;
  padding: 0 32px;
  -webkit-text-decoration: none;
  text-decoration: none;
  touch-action: manipulation;
  -webkit-transition: all 0.24s;
  transition: all 0.24s;
  width: auto;
}

.emotion-0:hover,
.emotion-0:focus {
  background-color: rgb(11,113,175);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
  >
    Submit
  </button>
</div>
`;

exports[`"Button" renders as block 1`] = `
.emotion-1 {
  background: rgb(247,248,250);
  -webkit-transition: background 0.15s;
  transition: background 0.15s;
}

.emotion-0 {
  color: #fff;
  display: inline-block;
  font-size: 16px;
  font-weight: 600;
  opacity: 1;
  overflow: hidden;
  text-align: center;
  text-overflow: ellipsis;
  text-transform: uppercase;
  white-space: nowrap;
  background-color: #0c78ba;
  border: 0;
  border-radius: 8px;
  cursor: pointer;
  height: 48px;
  line-height: 48px;
  margin-bottom: 16px;
  max-width: 100%;
  min-width: 150px;
  outline: none;
  padding: 0 32px;
  -webkit-text-decoration: none;
  text-decoration: none;
  touch-action: manipulation;
  -webkit-transition: all 0.24s;
  transition: all 0.24s;
  width: 100%;
}

.emotion-0:hover,
.emotion-0:focus {
  background-color: rgb(11,113,175);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
  >
    Submit
  </button>
</div>
`;

exports[`"Button" renders in loading state 1`] = `
@keyframes animation-0 {
  0% {
    stroke-dashoffset: 251.36;
  }

  100% {
    stroke-dashoffset: 0;
  }
}

@keyframes animation-1 {
  0% {
    -webkit-transform: rotate(0);
    -ms-transform: rotate(0);
    transform: rotate(0);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.emotion-2 {
  background: rgb(247,248,250);
  -webkit-transition: background 0.15s;
  transition: background 0.15s;
}

.emotion-1 {
  color: #fff;
  display: inline-block;
  font-size: 16px;
  font-weight: 600;
  opacity: 1;
  overflow: hidden;
  text-align: center;
  text-overflow: ellipsis;
  text-transform: uppercase;
  white-space: nowrap;
  background-color: #0c78ba;
  border: 0;
  border-radius: 8px;
  cursor: pointer;
  height: 48px;
  line-height: 48px;
  margin-bottom: 16px;
  max-width: 100%;
  min-width: 150px;
  outline: none;
  padding: 0 32px;
  -webkit-text-decoration: none;
  text-decoration: none;
  touch-action: manipulation;
  -webkit-transition: all 0.24s;
  transition: all 0.24s;
  width: auto;
}

.emotion-1:hover,
.emotion-1:focus {
  background-color: rgb(11,113,175);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0 {
  -webkit-animation: animation-0 2400ms ease-in-out infinite, animation-1 2400ms linear -1800ms infinite;
  animation: animation-0 2400ms ease-in-out infinite, animation-1 2400ms linear -1800ms infinite;
  stroke-dasharray: 125.68;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
}

<div
  className="emotion-2"
>
  <button
    className="emotion-1"
  >
    <svg
      fill="none"
      height={30}
      strokeWidth="3"
      style={
        Object {
          "display": "flex",
          "margin": "auto",
        }
      }
      viewBox="0 0 43 43"
      width={30}
    >
      <circle
        cx="21.5"
        cy="21.5"
        r="20"
        stroke="rgba(255, 255, 255, 0.07999999999999996)"
      />
      <circle
        className="emotion-0"
        cx="21.5"
        cy="21.5"
        r="20"
        stroke="#fff"
        strokeDasharray={125.68}
        strokeLinecap="round"
      />
    </svg>
  </button>
</div>
`;

exports[`"Button" renders when disabled 1`] = `
.emotion-1 {
  background: rgb(247,248,250);
  -webkit-transition: background 0.15s;
  transition: background 0.15s;
}

.emotion-0 {
  color: #fff;
  display: inline-block;
  font-size: 16px;
  font-weight: 600;
  opacity: 0.56;
  overflow: hidden;
  text-align: center;
  text-overflow: ellipsis;
  text-transform: uppercase;
  white-space: nowrap;
  background-color: #0c78ba;
  border: 0;
  border-radius: 8px;
  cursor: default;
  height: 48px;
  line-height: 48px;
  margin-bottom: 16px;
  max-width: 100%;
  min-width: 150px;
  outline: none;
  padding: 0 32px;
  -webkit-text-decoration: none;
  text-decoration: none;
  touch-action: manipulation;
  -webkit-transition: all 0.24s;
  transition: all 0.24s;
  width: auto;
}

.emotion-0:hover,
.emotion-0:focus {
  background-color: #0c78ba;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
    disabled={true}
  >
    Submit
  </button>
</div>
`;
